### DISCUSSION TITLE
Suggestion for copying lots of field values in fieldArray?

### DISCUSSION THREAD

**Discussion by Geddard** [react-hook-form/react-hook-form#4073] [Q&A]
Hey there, i have a situation where I have a fieldArray, with many values and some of them are more complex than usual (meaning they have nested properties).

I need to for example fill a row of horizontal fields, and then when i click a button i copy down these values to many rows below (with the same field structure). Sometimes incrementing some numeric values like "code0001" will get incremented by 1 for each extra row.

Is there a suggested way of handling bulk setting of values?
Right now im doing something like this:

```
(fields: string[], rowNumber: number) => { // fields are the names of fields we want to copy, and rowNumber is the reference to know where to start from to copy down
      const formRows = getValues()[FORM_NAME];
      let referenceRow = formRows[rowNumber];
      const newFormValues: Record<string, unknown>[] = [];

      formRows.forEach(() => {
        const newRow = {};

        fields.forEach(key => {
          const newValue = transformValue(key, get(referenceRow, key));

          set(newRow, key, newValue);
        });
        newFormValues.push(newRow);
        referenceRow = newRow;
      });

      reset({
        [FORM_NAME]: newFormValues,
      });
    },
    [getValues, reset, transformValue]
  );
```

It works but is underperforming heavily, it takes a long time for the calculation of the new fields and then the reset it's also quite slow (depending on the size of the array of course)

I think it was working better when I was using setValue but it doesn't seem to work with complex fields...

Suggestions?

Thanks!

**Comment by bluebill1049** [comment]
try this: `set('test', [] // deep nested field array value)` 

oh wait you already tried, what's not working? can you explain in detail?

**Reply by Geddard** [reply]
> try this: `set('test', [] // deep nested field array value)`
> 
> oh wait you already tried, what's not working? can you explain in detail?

What i was seeing was that the values were present in the final form object but the field listeners didn't seen to react to the updates so i was not seeing the values automatically populated.
Then i read in the Docs and here in some issues/questions that it was mentioned that setValue doesn't know how to handle fields with nested values like 

```
setValue(`${FORM_NAME}[${index}]`, valuesToCopy); 
```

 values to copy is the whole row, for example `{ code: { value: "asd", label: "asd"}, name:"asd", ...otherfields}`

**Reply by Geddard** [reply]
I guess the question is, does it work if i do `setValue('createForm[0]', objectWithNestedProps)`? or should i do it differently?

**Reply by bluebill1049** [reply]
```
setValue('createForm', [
    objectWithNestedProps,
   ...getValues('createForm).slice(1)
])
```

only at field name level right now.

**Reply by Geddard** [reply]
> ```
> setValue('createForm', [
>     objectWithNestedProps,
>    ...getValues('createForm).slice(1)
> ])
> ```
> 
> only at field name level right now.

So, if i understand correctly, i should pass the name of the form object (which contains the nested array) and then pass the whole array plus the new row each time?

**Reply by bluebill1049** [reply]
yes, correct.

**Reply by bluebill1049** [reply]
we should be able to improve this on V7.

**Reply by bluebill1049** [reply]
you can leave a comment there if you want: https://github.com/react-hook-form/react-hook-form/discussions/3714

**Reply by bluebill1049** [reply]
just double-checking aginst the code base, I think to keep at the array field key level is a lot easier to implement.

**Reply by Geddard** [reply]
So this will not work if i want to replace like `setValue('arrayName', newArray)` where new array has objects with nested props?

**Reply by Geddard** [reply]
Doing it the way you suggested now is a bit better, but still if i want to copy down something like 20 rows, its quite slow

**Reply by bluebill1049** [reply]
do you have a codesandbox?

**Reply by Geddard** [reply]
> do you have a codesandbox?

No, sorry, I have my hands quite full with the project and I'm juggling between handling the form and implementing some crazy fields here... What i have seen is that if i try to `setValue(arrayName, [nestedObjects])`, the following objects where for example code should be `code: { ...fields }` ends up just as `code: ""`

**Reply by Geddard** [reply]
The problem with using setValue for the whole row is that it's causing a full re-render of it, even the ids of the fieldArray change, so if i try to setValue to a form with up to 60 rows, it takes almost 10 seconds to re-render everything :/
The setValue for a field inside the array works like a charm, but it doesn't work for setting a value that is a complex object :(

**Reply by Geddard** [reply]
for example: 
```
setValue(`${FORM_NAME}[${index}][${key}]`, newValue);
```

**Reply by bluebill1049** [reply]
hmmm, looks like you will have to build a function update each inputs with the example you give above then...

**Reply by nani1008** [reply]
@Geddard @bluebill1049 any solution for this one? where i click a button and get the values from those input and auto populate it to my fieldArray?
